File: lib/tools/odds-api.ts
const BASE_URL = 'https://api.the-odds-api.com/v4'
const SPORT = 'soccer'

interface OddsOutcome { name: string; price: number; point?: number }
interface OddsMarket { key: string; outcomes: OddsOutcome[] }
interface OddsBookmaker { key: string; title: string; markets: OddsMarket[] }
interface OddsEvent {
  id: string
  sport_key: string
  sport_title: string
  home_team: string
  away_team: string
  commence_time: string
  bookmakers: OddsBookmaker[]
}

// Bookmaker prioritaire pour la cote finale — cohérence avec le lien affilié.
const PRIORITY_BOOKMAKER_KEY = '1xbet'

// Le sport_key 'soccer' interroge TOUTES les compétitions en une requête —
// pratique pour économiser le quota, mais ça inclut des championnats
// obscurs (réserves, divisions mineures de pays peu suivis) que les
// abonnés ne reconnaissent pas. Utilisé pour filtrer en mode "cotes
// uniquement" (lib/agents/analyst.ts), où il n'y a pas de données
// API-Football pour recadrer la sélection — même liste de championnats
// "connus" que lib/tools/flags.ts::COMPETITION_FLAG, pour rester cohérent.
const KNOWN_LEAGUE_SPORT_KEYS = [
  'epl', // Premier League (soccer_epl)
  'efl_champ', // Championship
  'la_liga', // La Liga
  'serie_a', // Serie A (Italie)
  'bundesliga', // Bundesliga
  'ligue_one', // Ligue 1
  'primeira_liga', // Liga Portugal
  'eredivisie', // Eredivisie
  'brazil_campeonato', // Campeonato Brasileiro
  'champs_league', // Ligue des Champions (UEFA + autres confédérations)
  'europa_league', // Europa League
  'conference_league', // Conference League
  'fifa_world_cup', // Coupe du monde
  'uefa_euro', // Euro (UEFA)
  'mls', // Major League Soccer
  'liga_mx', // Liga MX
]

export function isKnownLeague(sportKey: string): boolean {
  const key = sportKey.toLowerCase()
  return KNOWN_LEAGUE_SPORT_KEYS.some(k => key.includes(k))
}

export interface MatchOdds {
  sport_key: string
  sport_title: string
  home_team: string
  away_team: string
  commence_time: string
  h2h: { home: number | null; draw: number | null; away: number | null }
  totals: { over_2_5: number | null; under_2_5: number | null }
  btts: { yes: number | null; no: number | null }  // toujours null (btts non disponible en plan gratuit)
  // Handicap — seule ligne représentative (celle du premier bookmaker qui en
  // propose une), affichée à titre indicatif. La vérification multi-
  // bookmaker d'un pick handicap précis se fait via getBookmakerQuotes(),
  // qui filtre par valeur de point exacte, pas via ce champ.
  spreads: { home_point: number | null; home_price: number | null; away_point: number | null; away_price: number | null }
}

export async function getTodayOdds(region = 'eu'): Promise<MatchOdds[]> {
  const url = new URL(`${BASE_URL}/sports/${SPORT}/odds`)
  url.searchParams.set('apiKey', process.env.ODDS_API_KEY!)
  url.searchParams.set('regions', region)
  url.searchParams.set('markets', 'h2h,totals,spreads')  // btts retiré : non supporté plan gratuit
  url.searchParams.set('oddsFormat', 'decimal')
  url.searchParams.set('dateFormat', 'iso')

  const res = await fetch(url.toString(), { cache: 'no-store' })
  if (!res.ok) {
    console.warn(`Odds API ${res.status} — continuing without real-time odds`)
    return []
  }

  const events: OddsEvent[] = await res.json()

  return events.map(event => {
    const allMarkets = event.bookmakers.flatMap(b => b.markets)
    const h2hOutcomes = allMarkets.filter(m => m.key === 'h2h').flatMap(m => m.outcomes)
    const totalsOutcomes = allMarkets.filter(m => m.key === 'totals').flatMap(m => m.outcomes)
    const bttsOutcomes = allMarkets.filter(m => m.key === 'btts').flatMap(m => m.outcomes)
    const spreadsOutcomes = allMarkets.filter(m => m.key === 'spreads').flatMap(m => m.outcomes)

    const avgOdds = (outcomes: OddsOutcome[], name: string) => {
      const matching = outcomes.filter(o => o.name.toLowerCase().includes(name.toLowerCase()))
      if (!matching.length) return null
      return Math.round((matching.reduce((s, o) => s + o.price, 0) / matching.length) * 100) / 100
    }

    // Ligne représentative (premier bookmaker qui en propose une) — juste
    // pour informer l'Analyste qu'un handicap existe. La cote publiée vient
    // toujours de la vérification multi-bookmaker du Sélecteur de cotes.
    const homeSpread = spreadsOutcomes.find(o => o.name === event.home_team && o.point != null)
    const awaySpread = spreadsOutcomes.find(o => o.name === event.away_team && o.point != null)

    return {
      sport_key: event.sport_key,
      sport_title: event.sport_title,
      home_team: event.home_team,
      away_team: event.away_team,
      commence_time: event.commence_time,
      h2h: {
        home: avgOdds(h2hOutcomes, event.home_team),
        draw: avgOdds(h2hOutcomes, 'draw'),
        away: avgOdds(h2hOutcomes, event.away_team),
      },
      totals: {
        over_2_5: avgOdds(totalsOutcomes, 'over'),
        under_2_5: avgOdds(totalsOutcomes, 'under'),
      },
      btts: {
        yes: avgOdds(bttsOutcomes, 'yes'),
        no: avgOdds(bttsOutcomes, 'no'),
      },
      spreads: {
        home_point: homeSpread?.point ?? null,
        home_price: homeSpread?.price ?? null,
        away_point: awaySpread?.point ?? null,
        away_price: awaySpread?.price ?? null,
      },
    }
  })
}

// Cache en mémoire du process — évite de refaire un fetch par pick candidat
// quand l'Odds Selector interroge plusieurs matchs du même run.
let rawEventsCache: { data: OddsEvent[]; fetchedAt: number } | null = null
const RAW_CACHE_TTL_MS = 5 * 60 * 1000

async function fetchRawOddsEvents(region = 'eu'): Promise<OddsEvent[]> {
  if (rawEventsCache && Date.now() - rawEventsCache.fetchedAt < RAW_CACHE_TTL_MS) {
    return rawEventsCache.data
  }

  const url = new URL(`${BASE_URL}/sports/${SPORT}/odds`)
  url.searchParams.set('apiKey', process.env.ODDS_API_KEY!)
  url.searchParams.set('regions', region)
  url.searchParams.set('markets', 'h2h,totals,spreads')
  url.searchParams.set('oddsFormat', 'decimal')
  url.searchParams.set('dateFormat', 'iso')

  const res = await fetch(url.toString(), { cache: 'no-store' })
  if (!res.ok) {
    console.warn(`Odds API ${res.status} — cotes brutes indisponibles`)
    return []
  }

  const data = (await res.json()) as OddsEvent[]
  rawEventsCache = { data, fetchedAt: Date.now() }
  return data
}

// Similarité par mots communs (robuste aux abréviations et variantes de noms)
export function teamSimilarity(a: string, b: string): number {
  const normalize = (s: string) =>
    s.toLowerCase()
      .replace(/[^a-z0-9 ]/g, '')
      .replace(/\b(fc|cf|sc|rc|ac|ss|afc|bsc|vfb|rcd|ssc|ud|cd)\b/g, '')
      .trim()
  const wa = normalize(a).split(/\s+/).filter(Boolean)
  const wb = normalize(b).split(/\s+/).filter(Boolean)
  const setB = new Set(wb)
  const intersection = wa.filter(w => setB.has(w)).length
  const union = new Set([...wa, ...wb]).size
  return union === 0 ? 0 : intersection / union
}

export function findMatchOdds(
  odds: MatchOdds[],
  homeTeam: string,
  awayTeam: string
): MatchOdds | null {
  const THRESHOLD = 0.3 // au moins 1 mot commun significatif

  let best: MatchOdds | null = null
  let bestScore = -1

  for (const o of odds) {
    const homeScore = teamSimilarity(o.home_team, homeTeam)
    const awayScore = teamSimilarity(o.away_team, awayTeam)
    const score = homeScore + awayScore

    if (homeScore >= THRESHOLD && awayScore >= THRESHOLD && score > bestScore) {
      bestScore = score
      best = o
    }
  }

  return best
}

// ─── Cotes brutes par bookmaker (Odds Selector) ────────────────────────────

export interface BookmakerQuote {
  bookmaker: string // clé The Odds API (ex: '1xbet', 'pinnacle')
  price: number
}

// Reconnaît le marché/l'issue ciblés par un bet_type texte libre — même
// heuristique que lib/tools/result-checker.ts::evaluateResult (FR + EN, le
// Rédacteur/Analyste ne phrase pas toujours identiquement).
function matchOutcome(
  betType: string,
  homeTeam: string,
  awayTeam: string
): { marketKey: string; matches: (outcome: OddsOutcome) => boolean } | null {
  const bt = betType.toLowerCase()

  if (bt.includes('plus de 2.5') || bt.includes('over 2.5')) {
    return { marketKey: 'totals', matches: o => o.name.toLowerCase().includes('over') }
  }
  if (bt.includes('moins de 2.5') || bt.includes('under 2.5')) {
    return { marketKey: 'totals', matches: o => o.name.toLowerCase().includes('under') }
  }
  if (bt.includes('handicap')) {
    // "Handicap Real Madrid -1.5" — la ligne (point) doit matcher EXACTEMENT
    // celle du bookmaker (des lignes différentes ne sont pas comparables,
    // même erreur qu'on veut éviter que sur totals) et l'équipe visée.
    const pointMatch = betType.match(/[+-]?\d+(?:\.\d+)?/)
    const point = pointMatch ? parseFloat(pointMatch[0]) : null
    if (point == null) return null
    const targetTeam = teamSimilarity(betType, homeTeam) >= teamSimilarity(betType, awayTeam) ? homeTeam : awayTeam
    return {
      marketKey: 'spreads',
      matches: o => teamSimilarity(o.name, targetTeam) >= 0.3 && o.point != null && Math.abs(o.point - point) < 0.01,
    }
  }
  if (bt.includes('victoire') && (bt.includes('domicile') || bt.includes('home'))) {
    return { marketKey: 'h2h', matches: o => teamSimilarity(o.name, homeTeam) >= 0.3 }
  }
  if (bt.includes('victoire') && (bt.includes('extérieur') || bt.includes('exterieur') || bt.includes('away'))) {
    return { marketKey: 'h2h', matches: o => teamSimilarity(o.name, awayTeam) >= 0.3 }
  }
  if (bt.includes('nul') || bt.includes('draw')) {
    return { marketKey: 'h2h', matches: o => o.name.toLowerCase().includes('draw') }
  }

  // BTTS et autres marchés non couverts par le plan gratuit The Odds API
  return null
}

/**
 * Détail des cotes par bookmaker pour un pick candidat donné — utilisé par
 * l'Odds Selector pour choisir la cote 1xBet en priorité, sinon la médiane,
 * et détecter un marché trop incertain (écart entre bookmakers).
 */
export async function getBookmakerQuotes(homeTeam: string, awayTeam: string, betType: string): Promise<BookmakerQuote[]> {
  const outcome = matchOutcome(betType, homeTeam, awayTeam)
  if (!outcome) return []

  const events = await fetchRawOddsEvents()
  const THRESHOLD = 0.3
  const event = events.find(
    e => teamSimilarity(e.home_team, homeTeam) >= THRESHOLD && teamSimilarity(e.away_team, awayTeam) >= THRESHOLD
  )
  if (!event) return []

  const quotes: BookmakerQuote[] = []
  for (const bookmaker of event.bookmakers) {
    const market = bookmaker.markets.find(m => m.key === outcome.marketKey)
    const outcomeEntry = market?.outcomes.find(o => outcome.matches(o))
    if (outcomeEntry) quotes.push({ bookmaker: bookmaker.key, price: outcomeEntry.price })
  }
  return quotes
}

export { PRIORITY_BOOKMAKER_KEY }

File: backend/tools/odds_api.py
"""Port de lib/tools/odds-api.ts — The Odds API (api.the-odds-api.com/v4)."""
import os
import re
import time
from dataclasses import dataclass, field
from datetime import datetime, timedelta, timezone

import httpx

BASE_URL = "https://api.the-odds-api.com/v4"
SPORT = "soccer"

# Bookmaker prioritaire pour la cote finale — cohérence avec le lien affilié.
PRIORITY_BOOKMAKER_KEY = "1xbet"

# Même liste que côté site — utilisée pour filtrer en mode "cotes uniquement"
# (pas de données API-Football pour recadrer la sélection).
KNOWN_LEAGUE_SPORT_KEYS = [
    "epl", "efl_champ", "la_liga", "serie_a", "bundesliga", "ligue_one",
    "primeira_liga", "eredivisie", "brazil_campeonato", "champs_league",
    "europa_league", "conference_league", "fifa_world_cup", "uefa_euro",
    "mls", "liga_mx",
]


def is_known_league(sport_key: str) -> bool:
    key = sport_key.lower()
    return any(k in key for k in KNOWN_LEAGUE_SPORT_KEYS)


@dataclass
class TotalsLine:
    point: float
    over: float | None
    under: float | None


@dataclass
class MatchOdds:
    sport_key: str
    sport_title: str
    home_team: str
    away_team: str
    commence_time: str
    h2h: dict  # {home, draw, away}
    totals: list[TotalsLine] = field(default_factory=list)
    btts: dict = field(default_factory=lambda: {"yes": None, "no": None})
    spreads: dict = field(default_factory=lambda: {"home_point": None, "home_price": None, "away_point": None, "away_price": None})


def _day_window_utc(date: str) -> tuple[str, str]:
    """Fenêtre 00h-00h (UTC) pour une date donnée. Format EXACT attendu par
    l'API, sans les millisecondes (rejeté avec un 422 sinon — vérifié en
    conditions réelles)."""
    frm = f"{date}T00:00:00Z"
    to_dt = datetime.fromisoformat(frm.replace("Z", "+00:00")) + timedelta(days=1)
    to = to_dt.strftime("%Y-%m-%dT%H:%M:%SZ")
    return frm, to


async def get_today_odds(region: str = "eu", date: str | None = None) -> list[MatchOdds]:
    params = {
        "apiKey": os.environ["ODDS_API_KEY"],
        "regions": region,
        "markets": "h2h,totals,spreads",
        "oddsFormat": "decimal",
        "dateFormat": "iso",
    }
    if date:
        frm, to = _day_window_utc(date)
        params["commenceTimeFrom"] = frm
        params["commenceTimeTo"] = to

    async with httpx.AsyncClient() as client:
        res = await client.get(f"{BASE_URL}/sports/{SPORT}/odds", params=params, timeout=20.0)
        if not res.is_success:
            print(f"[warn] Odds API {res.status_code} — continuing without real-time odds")
            return []
        events = res.json()

    result = []
    for event in events:
        all_markets = [m for bm in event["bookmakers"] for m in bm["markets"]]
        h2h_outcomes = [o for m in all_markets if m["key"] == "h2h" for o in m["outcomes"]]
        totals_outcomes = [o for m in all_markets if m["key"] == "totals" for o in m["outcomes"]]
        btts_outcomes = [o for m in all_markets if m["key"] == "btts" for o in m["outcomes"]]
        spreads_outcomes = [o for m in all_markets if m["key"] == "spreads" for o in m["outcomes"]]

        def avg_odds(outcomes: list[dict], name: str) -> float | None:
            matching = [o for o in outcomes if name.lower() in o["name"].lower()]
            if not matching:
                return None
            return round((sum(o["price"] for o in matching) / len(matching)) * 100) / 100

        totals_points = sorted({o["point"] for o in totals_outcomes if o.get("point") is not None})
        totals = [
            TotalsLine(
                point=point,
                over=avg_odds([o for o in totals_outcomes if o.get("point") == point and "over" in o["name"].lower()], "over"),
                under=avg_odds([o for o in totals_outcomes if o.get("point") == point and "under" in o["name"].lower()], "under"),
            )
            for point in totals_points
        ]

        home_spread = next((o for o in spreads_outcomes if o["name"] == event["home_team"] and o.get("point") is not None), None)
        away_spread = next((o for o in spreads_outcomes if o["name"] == event["away_team"] and o.get("point") is not None), None)

        result.append(
            MatchOdds(
                sport_key=event["sport_key"],
                sport_title=event["sport_title"],
                home_team=event["home_team"],
                away_team=event["away_team"],
                commence_time=event["commence_time"],
                h2h={
                    "home": avg_odds(h2h_outcomes, event["home_team"]),
                    "draw": avg_odds(h2h_outcomes, "draw"),
                    "away": avg_odds(h2h_outcomes, event["away_team"]),
                },
                totals=totals,
                btts={"yes": avg_odds(btts_outcomes, "yes"), "no": avg_odds(btts_outcomes, "no")},
                spreads={
                    "home_point": home_spread["point"] if home_spread else None,
                    "home_price": home_spread["price"] if home_spread else None,
                    "away_point": away_spread["point"] if away_spread else None,
                    "away_price": away_spread["price"] if away_spread else None,
                },
            )
        )

    return result


# Cache en mémoire du process — évite de refaire un fetch par pick candidat
# quand l'Odds Selector interroge plusieurs matchs du même run.
_raw_events_cache: dict | None = None
_RAW_CACHE_TTL_SECONDS = 5 * 60


async def _fetch_raw_odds_events(region: str = "eu") -> list[dict]:
    global _raw_events_cache
    if _raw_events_cache and time.time() - _raw_events_cache["fetched_at"] < _RAW_CACHE_TTL_SECONDS:
        return _raw_events_cache["data"]

    params = {
        "apiKey": os.environ["ODDS_API_KEY"],
        "regions": region,
        "markets": "h2h,totals,spreads",
        "oddsFormat": "decimal",
        "dateFormat": "iso",
    }
    async with httpx.AsyncClient() as client:
        res = await client.get(f"{BASE_URL}/sports/{SPORT}/odds", params=params, timeout=20.0)
        if not res.is_success:
            print(f"[warn] Odds API {res.status_code} — cotes brutes indisponibles")
            return []
        data = res.json()

    _raw_events_cache = {"data": data, "fetched_at": time.time()}
    return data


_STOPWORDS = {"fc", "cf", "sc", "rc", "ac", "ss", "afc", "bsc", "vfb", "rcd", "ssc", "ud", "cd"}


def team_similarity(a: str, b: str) -> float:
    """Similarité par mots communs (robuste aux abréviations et variantes de noms)."""

    def normalize(s: str) -> list[str]:
        s = re.sub(r"[^a-z0-9 ]", "", s.lower())
        return [w for w in s.split() if w and w not in _STOPWORDS]

    wa, wb = normalize(a), set(normalize(b))
    if not wa and not wb:
        return 0.0
    intersection = len([w for w in wa if w in wb])
    union = len(set(wa) | wb)
    return intersection / union if union else 0.0


def find_match_odds(odds: list[MatchOdds], home_team: str, away_team: str) -> MatchOdds | None:
    threshold = 0.3
    best, best_score = None, -1.0

    for o in odds:
        home_score = team_similarity(o.home_team, home_team)
        away_score = team_similarity(o.away_team, away_team)
        score = home_score + away_score

        if home_score >= threshold and away_score >= threshold and score > best_score:
            best_score, best = score, o

    return best


@dataclass
class BookmakerQuote:
    bookmaker: str
    price: float


def _match_outcome(bet_type: str, home_team: str, away_team: str) -> dict | None:
    """Reconnaît le marché/l'issue ciblés par un bet_type texte libre — même
    heuristique que lib/tools/result-checker.ts::evaluate_result (FR + EN)."""
    bt = bet_type.lower()

    over_match = re.search(r"(?:plus de|over)\s*(\d+(?:\.\d+)?)", bet_type, re.IGNORECASE)
    if over_match:
        point = float(over_match.group(1))
        return {
            "market_key": "totals",
            "matches": lambda o: "over" in o["name"].lower() and o.get("point") is not None and abs(o["point"] - point) < 0.01,
        }

    under_match = re.search(r"(?:moins de|under)\s*(\d+(?:\.\d+)?)", bet_type, re.IGNORECASE)
    if under_match:
        point = float(under_match.group(1))
        return {
            "market_key": "totals",
            "matches": lambda o: "under" in o["name"].lower() and o.get("point") is not None and abs(o["point"] - point) < 0.01,
        }

    if "handicap" in bt:
        point_match = re.search(r"[+-]?\d+(?:\.\d+)?", bet_type)
        if not point_match:
            return None
        point = float(point_match.group(0))
        target_team = home_team if team_similarity(bet_type, home_team) >= team_similarity(bet_type, away_team) else away_team
        return {
            "market_key": "spreads",
            "matches": lambda o, _t=target_team: team_similarity(o["name"], _t) >= 0.3 and o.get("point") is not None and abs(o["point"] - point) < 0.01,
        }

    if "victoire" in bt and ("domicile" in bt or "home" in bt):
        return {"market_key": "h2h", "matches": lambda o: team_similarity(o["name"], home_team) >= 0.3}

    if "victoire" in bt and ("extérieur" in bt or "exterieur" in bt or "away" in bt):
        return {"market_key": "h2h", "matches": lambda o: team_similarity(o["name"], away_team) >= 0.3}

    if "nul" in bt or "draw" in bt:
        return {"market_key": "h2h", "matches": lambda o: "draw" in o["name"].lower()}

    return None


async def get_bookmaker_quotes(home_team: str, away_team: str, bet_type: str) -> list[BookmakerQuote]:
    """Détail des cotes par bookmaker pour un pick candidat donné — utilisé
    par l'Odds Selector pour choisir la cote 1xBet en priorité, sinon la
    médiane, et détecter un marché trop incertain (écart entre bookmakers)."""
    outcome = _match_outcome(bet_type, home_team, away_team)
    if not outcome:
        return []

    events = await _fetch_raw_odds_events()
    threshold = 0.3
    event = next(
        (e for e in events if team_similarity(e["home_team"], home_team) >= threshold and team_similarity(e["away_team"], away_team) >= threshold),
        None,
    )
    if not event:
        return []

    quotes = []
    for bookmaker in event["bookmakers"]:
        market = next((m for m in bookmaker["markets"] if m["key"] == outcome["market_key"]), None)
        if not market:
            continue
        outcome_entry = next((o for o in market["outcomes"] if outcome["matches"](o)), None)
        if outcome_entry:
            quotes.append(BookmakerQuote(bookmaker=bookmaker["key"], price=outcome_entry["price"]))

    return quotes
